include/nbl/builtin/hlsl/bxdf/fresnel.hlsl

```diff
@@ -17,131 +20,449 @@ namespace hlsl
+        retval.eta = eta.real();
+        retval.etak = eta.imag();
+        retval.etak2 = eta.imag()*eta.imag();
+        retval.clampedCosTheta = clampedCosTheta;
+        return retval;
+    }
+
+    T operator()()
+    {
+        const scalar_type cosTheta2 = clampedCosTheta * clampedCosTheta;
+        //const float sinTheta2 = 1.0 - cosTheta2;
+
+        const T etaLen2 = eta * eta + etak2;
+        assert(hlsl::all(etaLen2 > hlsl::promote<T>(hlsl::exp2<scalar_type>(-numeric_limits<scalar_type>::digits))));
+        const T etaCosTwice = eta * clampedCosTheta * 2.0f;
+
+        const T rs_common = etaLen2 + (T)(cosTheta2);
+        const T rs2 = (rs_common - etaCosTwice) / (rs_common + etaCosTwice);
+
+        const T rp_common = etaLen2 * cosTheta2 + (T)(1.0);
+        const T rp2 = (rp_common - etaCosTwice) / (rp_common + etaCosTwice);
+
+        return (rs2 + rp2) * 0.5f;
+    }
+
+    T eta;
+    T etak;
+    T etak2;
+    scalar_type clampedCosTheta;
+};
+
+template<typename T NBL_PRIMARY_REQUIRES(concepts::FloatingPointLikeVectorial<T>)
+struct Dielectric
+{
+    using scalar_type = typename vector_traits<T>::scalar_type;
+    using vector_type = T;
+
+    static Dielectric<T> create(NBL_CONST_REF_ARG(T) eta, scalar_type cosTheta)
+    {
+        Dielectric<T> retval;
+        scalar_type absCosTheta = hlsl::abs(cosTheta);
+        retval.orientedEta = OrientedEtas<T>::create(absCosTheta, eta);
+        retval.absCosTheta = absCosTheta;
+        return retval;
+    }
+
+    static T __call(NBL_CONST_REF_ARG(T) orientedEta2, scalar_type absCosTheta)
+    {
+        const scalar_type sinTheta2 = 1.0 - absCosTheta * absCosTheta;
+
+        // the max() clamping can handle TIR when orientedEta2<1.0
+        const T t0 = hlsl::sqrt<T>(hlsl::max<T>(orientedEta2 - sinTheta2, hlsl::promote<T>(0.0)));
+        const T rs = (hlsl::promote<T>(absCosTheta) - t0) / (hlsl::promote<T>(absCosTheta) + t0);
+
+        const T t2 = orientedEta2 * absCosTheta;
+        const T rp = (t0 - t2) / (t0 + t2);
+
+        return (rs * rs + rp * rp) * 0.5f;
+    }
+
+    T operator()()
+    {
+        return __call(orientedEta.value * orientedEta.value, absCosTheta);
+    }
+
+    OrientedEtas<T> orientedEta;
+    scalar_type absCosTheta;
+};
+
+template<typename T NBL_PRIMARY_REQUIRES(concepts::FloatingPointLikeVectorial<T>)
+struct DielectricFrontFaceOnly
+{
+    using scalar_type = typename vector_traits<T>::scalar_type;
+    using vector_type = T;
+
+    static DielectricFrontFaceOnly<T> create(NBL_CONST_REF_ARG(T) eta, scalar_type absCosTheta)
+    {
+        Dielectric<T> retval;
+        retval.orientedEta = OrientedEtas<T>::create(absCosTheta, eta);
+        retval.absCosTheta = hlsl::abs<T>(absCosTheta);
+        return retval;
+    }
+
+    T operator()()
+    {
+        return Dielectric<T>::__call(orientedEta.value * orientedEta.value, absCosTheta);
+    }
+
+    OrientedEtas<T> orientedEta;
+    scalar_type absCosTheta;
+};
```

Comment:
hmm seems there's even no need for this class, you just need to pass a differently clamped Abs and oriented eta and you're good